libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Trace.hs

```diff
@@ -0,0 +1,194 @@
+{-# LANGUAGE BangPatterns #-}
+{-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE StandaloneDeriving #-}
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Test.Cardano.Ledger.Generic.Trace where
+
+import Cardano.Binary (ToCBOR)
+import Cardano.Ledger.BaseTypes (Globals, mkTxIxPartial)
+import Cardano.Ledger.Coin (Coin (..))
+import Cardano.Ledger.Core (EraRule)
+import qualified Cardano.Ledger.Core as Core
+import Cardano.Ledger.Era (Era (Crypto))
+import Cardano.Ledger.Shelley.LedgerState (AccountState (..), DPState, LedgerState (..), UTxOState (..))
+import Cardano.Ledger.Shelley.Rules.Delegs (DelegsEnv)
+import Cardano.Ledger.Shelley.Rules.Ledger (LEDGER, LedgerEnv (..))
+import Cardano.Ledger.Shelley.Rules.Ledgers (LEDGERS, LedgersEnv (..))
+import Cardano.Ledger.Shelley.Rules.Utxo (UtxoEnv)
+import Cardano.Ledger.Shelley.TxBody (DCert)
+import Cardano.Ledger.TxIn (TxIn)
+import Control.Monad.Trans.Reader (ReaderT (..))
+import Control.State.Transition.Extended (Embed (..), STS (..))
+import Control.State.Transition.Trace.Generator.QuickCheck (HasTrace (..))
+import Data.Default.Class (Default)
+import Data.Functor.Identity (Identity (runIdentity))
+import Data.Sequence (Seq (..))
+import Data.Sequence.Strict (StrictSeq)
+import Data.Set (Set)
+import GHC.Records (HasField)
+import Test.Cardano.Ledger.Generic.GenState
+  ( GenEnv (..),
+    GenState (..),
+    getBlocksizeMax,
+    getReserves,
+    getSlot,
+    getTreasury,
+  )
+-- import Control.State.Transition.Trace(applySTSTest,checkTrace)
+-- import Test.Cardano.Ledger.Shelley.Rules.Chain (CHAIN, ChainState (..), initialShelleyState)
+-- import qualified Cardano.Ledger.Babbage.PParams(PParams'(..))
+-- import qualified Cardano.Ledger.Alonzo.PParams(PParams'(..))
+-- import qualified Cardano.Ledger.Shelley.PParams(PParams'(..))
```

Comment:
Dead code?
```suggestion
```

Reply:
@lehins has the sixth sense of seeing dead code :laughing:

Reply:
This is a bit tricky, when developing code, and running examples the constructors of the PParams' types must be in scope or the HasField instance are not known, which means no example will run. I promise to delete this when all is done.

Reply:
all gone.